src/logic/hooks/useAsync.ts

```diff
@@ -3,32 +3,41 @@ import { useEffect, useState } from 'react'
 type AsyncResult<T> = {
   error: Error | undefined
   result: T | undefined
+  isLoading: boolean
 }
 
 const useAsync = <T>(asyncCall: () => Promise<T>): AsyncResult<T> => {
-  const [asyncVal, setAsyncVal] = useState<T>()
-  const [err, setErr] = useState<Error>()
+  const [result, setResult] = useState<T>()
+  const [error, setError] = useState<Error>()
+  const [isLoading, setIsLoading] = useState<boolean>(false)
 
   useEffect(() => {
     let isCurrent = true
 
-    setAsyncVal(undefined)
-    setErr(undefined)
+    setResult(undefined)
+    setError(undefined)
+    setIsLoading(true)
 
     asyncCall()
       .then((val: T) => {
-        if (isCurrent) setAsyncVal(val)
+        if (isCurrent) {
+          setResult(val)
+          setIsLoading(false)
+        }
       })
       .catch((error) => {
-        if (isCurrent) setErr(error)
+        if (isCurrent) {
+          setError(error)
+          setIsLoading(false)
+        }
       })
 
     return () => {
       isCurrent = false
     }
   }, [asyncCall])
 
-  return { error: err, result: asyncVal }
+  return { error, result, isLoading }
```

Comment:
It might be possible to derive `isLoading` from `error` and `result` i.e. `isLoading: !result && !error` so we wouldn't need an additional state for it.

Reply:
That was my first approach, but if the Promise gets rejected with an `undefined` value then you get a `isLoading` `true` value. I think that the best option here it is to add this state to make sure that we have a 100% generic hook.

```
  it('isLoading error value', async () => {
    const fakeCallback = jest.fn(() => Promise.reject())

    const { result } = renderHook(() => useAsync(fakeCallback))

    expect(fakeCallback).toHaveBeenCalledTimes(1)

    expect(result.current.result).toBe(undefined)
    expect(result.current.error).toBe(undefined)
    expect(result.current.isLoading).toBe(false) // Error here, Received: true
  })
```

Reply:
Ok, that makes sense if we want to store an error message instead of a boolean. Thanks for clarifying!